feat: add hiddenSourceMaps option to prevent sourcemap exposure

omareltomy opened the pull request:
- Add hiddenSourceMaps boolean option to Next.js config
- When enabled with productionBrowserSourceMaps, generates 'hidden-source-map' instead of 'source-map'
- Prevents # sourceMappingURL comments in generated JS bundles
- Allows error tracking services to use source maps while hiding them from end users
- Fixes S3/CDN 404 errors when source maps aren't uploaded publicly

Closes #[84095]

<!-- Thanks for opening a PR! Your contribution is much appreciated.
To make sure your PR is handled as smoothly as possible we request that you follow the checklist sections below.
Choose the right checklist for the change(s) that you're making:

## For Contributors

### Improving Documentation

- Run `pnpm prettier-fix` to fix formatting issues before opening the PR.
- Read the Docs Contribution Guide to ensure your contribution follows the docs guidelines: https://nextjs.org/docs/community/contribution-guide

### Fixing a bug

- Related issues linked using `fixes #number`
- Tests added. See: https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs
- Errors have a helpful link attached, see https://github.com/vercel/next.js/blob/canary/contributing.md

### Adding a feature

- Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR. (A discussion must be opened, see https://github.com/vercel/next.js/discussions/new?category=ideas)
- Related issues/discussions are linked using `fixes #number`
- e2e tests added (https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs)
- Documentation added
- Telemetry added. In case of a feature if it's used or not.
- Errors have a helpful link attached, see https://github.com/vercel/next.js/blob/canary/contributing.md


## For Maintainers

- Minimal description (aim for explaining to someone not on the team to understand the PR)
- When linking to a Slack thread, you might want to share details of the conclusion
- Link both the Linear (Fixes NEXT-xxx) and the GitHub issues
- Add review comments if necessary to explain to the reviewer the logic behind a change

## What?

Adds a `hiddenSourceMaps` config option that generates source maps without exposing them to users in production.

## Why?

When you enable `productionBrowserSourceMaps`, Next.js includes `# sourceMappingURL` comments in your JS bundles. This means:
- Users can access your source code through browser dev tools
- Your CDN gets 404 requests for source map files you don't want to upload publicly
- Error tracking works fine, but you're sharing more than you intended

## How?

The new option works alongside `productionBrowserSourceMaps`:

```js
// next.config.js
module.exports = {
  productionBrowserSourceMaps: true,
  hiddenSourceMaps: true, // no more sourceMappingURL comments
}

Closes #84095

-->

omareltomy committed:
feat: add hiddenSourceMaps option to prevent sourcemap exposure

- Add hiddenSourceMaps boolean option to Next.js config
- When enabled with productionBrowserSourceMaps, generates 'hidden-source-map' instead of 'source-map'
- Prevents # sourceMappingURL comments in generated JS bundles
- Allows error tracking services to use source maps while hiding them from end users
- Fixes S3/CDN 404 errors when source maps aren't uploaded publicly

Closes #[84095]

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 68f88cf4d8566aab81cf2fd526017d7dd904cda2

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

omareltomy committed:
fix: ensure hiddenSourceMaps only applies when productionBrowserSourceMaps is enabled

omareltomy committed:
Merge branch 'canary' into feat/hidden-sourcemaps-option

omareltomy reviewed (commented):
now you can't have "hidden source maps" if you're not generating source maps in the first place!

omareltomy committed:
fix: apply DevToolsIgnorePlugin to hidden-source-map as well

DevToolsIgnorePlugin should be applied to both 'source-map' and
'hidden-source-map' since they both generate actual source map files
that benefit from ignore list functionality for debugging.

holtkampjs commented:
@omareltomy 
It's sad to see this PR hasn't gotten more love. I've recently run into this as an issue.

Looking at Next.js's docs,

> Source Maps are enabled by default during development. During production builds, they are disabled to prevent you leaking your source on the client, unless you specifically opt-in with the configuration flag.

It seems clear that even the people who wrote the docs recognize that it's not often ideal to serve your sourcemaps unless that is something you explicitly want to do. However, there are plenty of users who want to generate sourcemaps for use cases such as error log de-obfuscation in logging and monitoring software.
